src/DocumentDB.ChangeFeedProcessor/PartitionManagement/RemainingWorkEstimator.cs

```diff
@@ -19,69 +22,73 @@ internal class RemainingWorkEstimator : IRemainingWorkEstimator
+            this.degreeOfParallelism = degreeOfParallelism;
         }
 
         public async Task<long> GetEstimatedRemainingWork()
         {
-            long remainingWork = 0;
-            bool hasAtLeastOneLease = false;
+            var partitions = await this.GetEstimatedRemainingWorkPerPartition();
+            if (partitions.Count == 0) return 1;
 
-            ChangeFeedOptions options = new ChangeFeedOptions
-            {
-                MaxItemCount = 1,
-            };
+            return partitions.Sum(partition => partition.RemainingWork);
+        }
 
-            foreach (ILease existingLease in await this.leaseManager.ListAllLeasesAsync().ConfigureAwait(false))
+        public async Task<IReadOnlyList<RemainingPartitionWork>> GetEstimatedRemainingWorkPerPartition()
+        {
+            IReadOnlyList<ILease> leases = await this.leaseManager.ListAllLeasesAsync().ConfigureAwait(false);
+            if (leases == null || leases.Count == 0)
             {
-                hasAtLeastOneLease = true;
-                options.PartitionKeyRangeId = existingLease.PartitionId;
-                options.RequestContinuation = existingLease.ContinuationToken;
-                options.StartFromBeginning = string.IsNullOrEmpty(existingLease.ContinuationToken);
-
-                IChangeFeedDocumentQuery<Document> query = this.feedDocumentClient.CreateDocumentChangeFeedQuery(this.collectionSelfLink, options);
-                IFeedResponse<Document> response = null;
+                return new List<RemainingPartitionWork>().AsReadOnly();
+            }
 
-                try
+            var tasks = Partitioner.Create(leases)
+                .GetPartitions(this.degreeOfParallelism)
+                .Select(partition => Task.Run(async () =>
                 {
-                    response = await query.ExecuteNextAsync<Document>().ConfigureAwait(false);
-                    long parsedLSNFromSessionToken = TryConvertToNumber(ExtractLSNFromSessionToken(response.SessionToken));
-                    long lastQueryLSN = response.Count > 0 ?
-                        TryConvertToNumber(GetFirstDocument(response).GetPropertyValue<string>(LSNPropertyName)) - 1
-                        : parsedLSNFromSessionToken;
-                    if (lastQueryLSN < 0)
+                    IList<RemainingPartitionWork> results = new List<RemainingPartitionWork>();
+                    using (partition)
                     {
-                        // Could not parse LSN from document, we cannot determine the amount of changes but since the query returned 1 document, we know it's at least 1
-                        remainingWork += 1;
-                        continue;
+                        while (partition.MoveNext())
+                        {
+                            ILease item = partition.Current;
+                            try
+                            {
+                                if (string.IsNullOrEmpty(item.PartitionId)) continue;
```

Comment:
yes. because leaseManager.ListAllLeasesAsync() will read all leases, including! "info" document.
e.g.
{
  "id": ".............._ErseAA==_ErseALG6egA=.info"
}

I wanted to discuss this with you:  

I dont like this in the code: 
Debug.Assert(lease.PartitionId != null, "TakeLeasesAsync: lease.PartitionId cannot be null.");

The besr way would be to change ListAllLeases() to return all but filter out "non partition leases", wdyt?

Reply:
yes, .info shouldn't be there